src/util/macro_util.rs

```diff
@@ -733,7 +674,183 @@ where
+    /// # PME
+    ///
+    /// Instantiating this method PMEs unless both:
+    /// - `mem::size_of::<Dst>() == mem::size_of::<Src>()`
+    /// - `mem::align_of::<Dst>() <= mem::align_of::<Src>()`
+    #[inline(always)]
+    #[must_use]
+    pub fn transmute_mut(self) -> &'a mut Dst
+    where
+        Src: FromBytes + IntoBytes,
+        Dst: FromBytes + IntoBytes,
+    {
+        static_assert!(Src, Dst => mem::size_of::<Dst>() == mem::size_of::<Src>());
+        static_assert!(Src, Dst => mem::align_of::<Dst>() <= mem::align_of::<Src>());
+
+        let src: *mut Src = self.0;
+        let dst = src.cast::<Dst>();
+        // SAFETY:
+        // - We know that it is sound to view the target type of the input
+        //   reference (`Src`) as the target type of the output reference
+        //   (`Dst`) and vice-versa because `Src: FromBytes + IntoBytes`, `Dst:
+        //   FromBytes + IntoBytes`, and (as asserted above) `size_of::<Src>()
+        //   == size_of::<Dst>()`.
+        // - We asserted above that alignment will not increase.
+        // - We know that the returned lifetime will not outlive the input
+        //   lifetime thanks to the lifetime bounds on this function.
+        unsafe { &mut *dst }
+    }
+}
+
+pub trait TransmuteRefDst<'a> {
+    type Dst: ?Sized;
+
+    #[must_use]
+    fn transmute_ref(self) -> &'a Self::Dst;
+}
+
+impl<'a, Src: ?Sized, Dst: ?Sized> TransmuteRefDst<'a> for Wrap<&'a Src, &'a Dst>
+where
+    Src: KnownLayout<PointerMetadata = usize> + IntoBytes + Immutable,
+    Dst: KnownLayout<PointerMetadata = usize> + FromBytes + Immutable,
+{
+    type Dst = Dst;
+
+    #[inline(always)]
+    fn transmute_ref(self) -> &'a Dst {
+        static_assert!(Src: ?Sized + KnownLayout, Dst: ?Sized + KnownLayout => {
+            Src::LAYOUT.align.get() >= Dst::LAYOUT.align.get()
+        }, "cannot transmute reference when destination type has higher alignment than source type");
+
+        safety_comment! {
+            /// SAFETY:
+            /// We only use `S` as `S<Src>` and `D` as `D<Dst>`.
+            unsafe_with_size_eq!(<S<Src>, D<Dst>> {
+                let ptr = Ptr::from_ref(self.0)
+                    .transmute::<S<Src>, invariant::Valid, BecauseImmutable>()
+                    .recall_validity::<invariant::Initialized, _>()
+                    .transmute::<D<Dst>, invariant::Initialized, (crate::pointer::BecauseMutationCompatible, _)>()
+                    .recall_validity::<invariant::Valid, _>();
+
+                // SAFETY: The preceding `static_assert!` ensures that
+                // `T::LAYOUT.align >= U::LAYOUT.align`. Since `self.0` is
+                // validly-aligned for `T`, it is also validly-aligned for `U`.
+                let ptr = unsafe { ptr.assume_alignment() };
+
+                &ptr.as_ref().0
+            })
+        }
+    }
+}
+
+pub trait TransmuteMutDst<'a> {
+    type Dst: ?Sized;
+    #[must_use]
+    fn transmute_mut(self) -> &'a mut Self::Dst;
+}
```

Comment:
Can we collapse `TransmuteRefDst` and `TransmuteMutDst` into a single trait?

Reply:
This doesn't work - we have to specifically implement for `Wrap<&Src>` and `Wrap<&mut Src>` separately - these can't be unified.